### ISSUE TITLE
Generated Instructions: derive behavioral guidelines from active ESLint config for AI coding tools

### ISSUE THREAD

**Issue by pertrai1** [pertrai1/eslint-plugin-llm-core#120]
## Summary

Ship a CLI command that reads the project's active `eslint-plugin-llm-core` config and generates a behavioral guidelines file (`.agents/linting-rules.md`) that AI coding tools can consume. Provider-specific instruction files reference or embed the generated content.

**Phased delivery.** This issue covers Phase 1 (core generation). Provider adapters are Phase 2.

## Motivation

Lint rules are reactive — they fire after code is written. Generated instructions are proactive — they encode the same rule knowledge as imperative behavioral guidelines that shape LLM output before violations occur.

Unlike static skills, generated instructions are:
1. **Personalized** — derived from the user's actual rule config and option values
2. **In sync** — regenerated when config changes, no drift between rules and instructions
3. **File-type-aware** — TS-only rules appear in a separate section, not as universal guidance

No other ESLint plugin does this.

## Relationship to Skills

Generated instructions **replace rule-derived skills** (like the proposed `code-reviewer` from #87, which was just restating lint rules as static markdown). They do **NOT** replace judgment-based skills like `test-reviewer`, which covers heuristics that cannot be expressed as lint rules ("don't mirror implementation logic in tests," "assert behavior not mocks"). That guidance is a different artifact — keep `skills/test-reviewer.md`.

Closes #87 on completion.

### Phase 1: Core Generation (this issue)

Ship `.agents/linting-rules.md` generation. Prove the rule metadata model and config resolution logic.

### Phase 2: Provider Adapters (follow-up issue)

Add provider-specific file management — reference-based adapters (Claude Code, OpenCode) and embedded adapters (Copilot). CLI interactive mode with provider selection, `--update` for regeneration, marker-based embedding.

---

### 1. Rule instruction metadata

Each rule gets structured instruction metadata co-located with its rule definition. This is a **separate artifact** from lint messages — lint messages are post-violation repair (before/after code rewrites), instructions are pre-write principles.

**Metadata structure per rule:**
```typescript
instruction: {
  principle: string;          // Imperative behavioral guideline with {option} placeholders
  optionTemplate?: string;    // Override when principle needs different rendering with options
}
```

**No `scope` field.** Scope is derived by the generator from resolved effective config (see section 2), not maintained manually. The rule author should not need to keep scope in sync with config composition — that's the generator's job.

**Examples:**

```typescript
// no-exported-function-expressions
instruction: {
  principle: "Use function declarations for all exports — `export function foo()` not `export const foo = () =>`",
}

// max-complexity (single option → {max} placeholder in principle)
instruction: {
  principle: "Keep cyclomatic complexity under {max} — decompose when functions get complex",
}

// max-file-length (single option → {max} placeholder)
instruction: {
  principle: "Keep files under {max} lines — split modules when they exceed this",
}

// no-magic-numbers (multiple options — see option handling rules below)
instruction: {
  principle: "Extract named constants for all magic numbers",
  optionTemplate: "Extract named constants for magic numbers (ignore: {ignoreNumbers})",
}
```

**Option interpolation rules:**

| Case | Behavior | Example |
|------|----------|---------|
| No options | Use `principle` as-is | `no-empty-catch` → "Never leave catch blocks empty" |
| Single numeric option | Replace `{optionName}` in `principle` | `max-complexity` max:8 → "under 8" |
| Single boolean option | Append "enabled" / omit | `no-commented-out-code` — no options to interpolate |
| Multiple options | Use `optionTemplate` if present; otherwise use `principle` without interpolation | `no-magic-numbers` with `ignore` array → use `optionTemplate` |
| Array options | Render as comma-separated list | `ignore: [0, 1]` → "0, 1" |

**Deterministic rule:** If a rule has options that affect behavioral guidance, the author provides either `{placeholders}` in `principle` (for simple cases) or a separate `optionTemplate` (for complex cases). The generator interpolates values at generation time. If neither is provided, the principle is rendered as-is — no fallback interpolation.

### 2. Config resolution

The generator reads the user's `eslint.config.*` and resolves which llm-core rules are active (not `off`) and their configured option values. This uses ESLint's config resolution API — not manual parsing.

**Scope is derived, not declared.** The generator resolves effective config for two virtual file paths:
- `path.join(process.cwd(), '__virtual__.js')`
- `path.join(process.cwd(), '__virtual__.ts')`

ESLint's flat config matches rules to files by resolving glob patterns against **absolute paths**. A bare `virtual.js` won't match a user's `**/*.js` glob — the virtual paths must be fully resolved relative to the project root. Using `process.cwd()` ensures the paths match the same resolution context ESLint uses during normal linting.

A rule's scope is determined by where it appears:
- Present in both → "All Files" section
- Present for `.ts` only → "TypeScript Files Only" section
- Present for `.js` only → "All Files" section (edge case, unlikely but handled)

Rule authors never maintain a `scope` field. Scope emerges from the project's actual config composition.

**Severity rendering:** Rules set to `"warn"` and `"error"` render identically in generated output. There is no distinction — both produce the same imperative guideline. Phase 1 treats them uniformly; if softening warn-level guidance proves valuable later, that's a Phase 2 decision with its own acceptance criteria.

### 3. Generated file: `.agents/linting-rules.md`

```markdown
# Coding Guidelines

Generated from eslint-plugin-llm-core configuration.
Regenerate with: npx eslint-plugin-llm-core generate-instructions

## All Files

- Use function declarations for all exports — `export function foo()` not `export const foo = () =>`
- Use structured logging with static message strings; pass dynamic values as separate metadata
- Never leave catch blocks empty — handle, rethrow, or log the error
- Always throw Error objects, never strings or plain objects
- Keep cyclomatic complexity under 8 — decompose when functions get complex
- Keep functions under 50 lines — extract helpers when they grow
- Prefer early returns over wrapping bodies in if statements
- Extract named constants for all magic numbers
- No commented-out code, no eslint-disable comments, no placeholder comments

## TypeScript Files Only

- Add explicit parameter and return type annotations on all exported functions
- Never use `any` as a generic type argument
- Never use type assertions to `any`
- Use `unknown` for catch parameter types, not `any`
```

Rules that are `off` do not appear. Configured option values are embedded (e.g., "under 8" because `max: 8`, not generic "keep complexity low"). If all active rules apply to both file types, the "TypeScript Files Only" section is omitted entirely.

### 4. CLI command

```bash
npx eslint-plugin-llm-core generate-instructions
```

Reads `eslint.config.*` from CWD, resolves active rules, generates `.agents/linting-rules.md`. If the file already exists, overwrites it (the header makes clear it's auto-generated).

Flags:
- `--config <path>` — specify config file path (default: auto-detect)
- `--dry-run` — print generated content to stdout without writing files

No interactive prompts in Phase 1. Provider selection and multi-file management come in Phase 2.

### 5. Programmatic export

```typescript
import { generateInstructions } from "eslint-plugin-llm-core/instructions";

const result = await generateInstructions({
  configPath: "eslint.config.mjs",
});

result.content;          // The generated markdown string
result.activeRules;      // Resolved rules with options and resolved scope
result.allFilesRules;    // Rules applying to all files
result.typescriptRules;  // TS-only rules
```

**Implementation surface:** The separate export path and CLI require updates to `package.json`:
- New entry in `exports` map for `eslint-plugin-llm-core/instructions`
- New entry in `bin` field pointing to the compiled CLI entry point (e.g., `"generate-instructions": "./dist/cli/generate-instructions.js"`)
- Corresponding build output in `dist/instructions.*` and `dist/cli/generate-instructions.js`
- TypeScript compilation config adjustment if the generator or CLI source lives outside `src/`

This is part of the implementation, not an afterthought. The build pipeline must produce the main plugin bundle, the instructions module, and the CLI entry point.

## Phase 1 Acceptance Criteria

- [ ] All 24 current rules have instruction metadata (`principle`, optional `optionTemplate`) co-located with rule definitions
- [ ] Instruction metadata does NOT include a `scope` field — scope is derived by the generator from resolved config
- [ ] `npx eslint-plugin-llm-core generate-instructions` creates `.agents/linting-rules.md`
- [ ] CLI entry point is registered in `package.json` `bin` field
- [ ] Only rules set to `error` or `warn` appear; `off` rules are excluded
- [ ] Rules at `warn` and `error` render identically in generated output
- [ ] Configured option values are interpolated into generated text per the option handling rules
- [ ] Virtual file paths use `path.join(process.cwd(), '__virtual__.{js,ts}')` for correct glob matching
- [ ] TS-only rules appear in a separate "TypeScript Files Only" section (derived from config resolution, not manual scope)
- [ ] If no TS-only rules are active, the "TypeScript Files Only" section is omitted
- [ ] Rules applying to all files appear in "All Files" section
- [ ] `--config` flag allows specifying config file path
- [ ] `--dry-run` flag prints to stdout without writing files
- [ ] Programmatic `generateInstructions` export works from `eslint-plugin-llm-core/instructions`
- [ ] `package.json` exports map includes the `instructions` entry point
- [ ] `package.json` bin field includes the CLI entry point
- [ ] Build pipeline produces main plugin, instructions module, and CLI entry point
- [ ] `tsc --noEmit` passes
- [ ] `npm test` passes (including tests for generator, config resolution, option interpolation, and metadata)
- [ ] `npm run lint` passes
- [ ] `npm run build` passes
- [ ] README updated with "Generated Instructions" section
- [ ] `skills/test-reviewer.md` is preserved (NOT deleted)
- [ ] `skills/code-reviewer.md` is NOT created (replaced by this feature)
- [ ] Closes #87

## Out of Scope (Phase 1)

- Provider file management (`.cursorrules`, `copilot-instructions.md`, etc.) — Phase 2
- Interactive provider selection — Phase 2
- `--update` mode with saved provider cache — Phase 2
- Marker-based embedding in existing files — Phase 2
- Instructions for non-llm-core rules — future consideration
- MCP server or runtime drift detection — different problem, different solution
- Removing `skills/test-reviewer.md` — it covers judgment-based heuristics, not rule restatements
- Differentiated rendering for `warn` vs `error` severity — evaluate after Phase 1 ships

## Evaluation Criteria (not acceptance criteria)

Do repos using generated instructions produce fewer `llm-core` violations before lint runs? If yes, this becomes the plugin's clearest differentiator. This is measured after shipping, not gated on it.
